File: src/Organon/Syl/Validity.hs
module Organon.Syl.Validity
  ( ValidationResult (..),
    validate,
    identifyMood,
  )
where

import Data.List (find)
import Data.Text (Text)
import Organon.Syl.Tradition
import Organon.Syl.Types

-- | The result of validating a syllogism.
data ValidationResult
  = -- | Valid with the given mood, premises in the supplied order.
    Valid Mood
  | -- | Valid with the given mood, but only after swapping premises.
    ValidSwapped Mood Syllogism
  | Invalid Text
  deriving stock (Eq, Show)

-- | Identify the mood of a syllogism, if it matches any known mood
-- in the given tradition.
identifyMood :: Tradition -> Syllogism -> Maybe Mood
identifyMood tradition syl =
  case figure syl of
    Nothing -> Nothing
    Just fig ->
      let pt = (propType (major syl), propType (minor syl), propType (conclusion syl))
       in findMood fig pt (validMoods tradition)

-- | Validate a syllogism against a tradition.
-- Tries both premise orderings: if the original fails figure
-- detection, swaps the premises and retries.
validate :: Tradition -> Syllogism -> ValidationResult
validate tradition syl =
  case identifyMood tradition syl of
    Just m -> Valid m
    Nothing ->
      let swapped = syl {major = minor syl, minor = major syl}
       in case identifyMood tradition swapped of
            Just m -> ValidSwapped m swapped
            Nothing ->
              case (figure syl, figure swapped) of
                (Nothing, Nothing) ->
                  Invalid "Propositions do not form a well-structured syllogism (no consistent middle term)"
                _ ->
                  Invalid "Not a valid syllogism in the given tradition"

findMood :: Figure -> (PropType, PropType, PropType) -> [Mood] -> Maybe Mood
findMood fig pt = find $ \m ->
  let spec = moodSpec m
   in moodFigure spec == fig
        && (majorPropType spec, minorPropType spec, conclusionPropType spec) == pt

File: web/src/core/validity.ts
// Syllogism validation — ported from Organon.Syl.Validity

import { Figure, Mood, PropType, Syllogism, Tradition, figure } from "./types";
import { moodSpec, validMoods } from "./tradition";

export type ValidationResult =
  | { readonly tag: "Valid"; readonly mood: Mood }
  | {
      readonly tag: "ValidSwapped";
      readonly mood: Mood;
      readonly syllogism: Syllogism;
    }
  | { readonly tag: "Invalid"; readonly message: string };

export function identifyMood(
  tradition: Tradition,
  syl: Syllogism,
): Mood | null {
  const fig = figure(syl);
  if (fig === null) return null;
  const pt: [PropType, PropType, PropType] = [
    syl.major.propType,
    syl.minor.propType,
    syl.conclusion.propType,
  ];
  return findMood(fig, pt, validMoods(tradition));
}

export function validate(
  tradition: Tradition,
  syl: Syllogism,
): ValidationResult {
  const m = identifyMood(tradition, syl);
  if (m !== null) return { tag: "Valid", mood: m };

  const swapped: Syllogism = {
    major: syl.minor,
    minor: syl.major,
    conclusion: syl.conclusion,
  };
  const ms = identifyMood(tradition, swapped);
  if (ms !== null) return { tag: "ValidSwapped", mood: ms, syllogism: swapped };

  const fig1 = figure(syl);
  const fig2 = figure(swapped);
  if (fig1 === null && fig2 === null) {
    return {
      tag: "Invalid",
      message:
        "Propositions do not form a well-structured syllogism (no consistent middle term)",
    };
  }
  return {
    tag: "Invalid",
    message: "Not a valid syllogism in the given tradition",
  };
}

function findMood(
  fig: Figure,
  pt: [PropType, PropType, PropType],
  moods: Mood[],
): Mood | null {
  for (const m of moods) {
    const spec = moodSpec(m);
    if (
      spec.moodFigure === fig &&
      spec.majorPropType === pt[0] &&
      spec.minorPropType === pt[1] &&
      spec.conclusionPropType === pt[2]
    ) {
      return m;
    }
  }
  return null;
}
